### PR TITLE
Fixed #24639 -- ValidationError respects SafeData.

### PR DESCRIPTION
**Pull request by jambonrose**
This is a fix for [ticket 24639](https://code.djangoproject.com/ticket/24639), provided that the state table in the ticket is actually what we desire.

### CODE REVIEW

File: django/core/exceptions.py [hunk 1]
```diff
@@ -162,6 +163,9 @@ def __iter__(self):
                 message = error.message
                 if error.params:
                     message %= error.params
+                    if (isinstance(error.message, SafeData) and
+                        all([isinstance(v, SafeData) for v in error.params.values()])):
```
Comment (charettes) [inline django/core/exceptions.py]: Unnecessary list comprehension, `all(isinstance(v, SafeDate) ...)` should do.

Comment (jambonrose) [inline django/core/exceptions.py]: Very cool. Fixed.

**Comment by timgraham** [comment]
buildbot, add to whitelist

**Comment by timgraham** [comment]
Conceptually this makes sense to me. I guess we'd run into a problem if two strings that are marked safe  interpolate in a way that becomes unsafe, but it seems like that's a case the developer should account for when marking strings as safe.

@loic, since you did a lot of work on error messages, do you have any feedback?

**Comment by loic** [comment]
IMO `ValidationError` should be HTML agnostic since it's the responsibility of `ErrorList/Dict` to deal with the interaction with HTML. This also seems pretty fragile, potentially unused params dictate the final outcome and so do safe params such as `int` and `float`. HTML in Python code is not something we want to promote anyway (HTML belongs in templates). The logic behind using `params` is to make things as atomic as possible to give more flexibility to the user of the form. If you are going to couple two different technologies, you might as well interpolate the final message yourself.

If we really wanted to deal with interpolation/formatting of `SafeData` it should be addressed using `__format__` and `__mod__` on the `SafeData` object itself.

**Comment by loic** [comment]
It's also worth noting that this implementation assumes that `params` is a `dict`. We recommend using a `dict` since Django 1.8 since it gives more control over formatting, but it's pretty common for people to use a `list` or `tuple` instead.

**Comment by timgraham** [comment]
So your vote would be a "won't fix" then? In that case should we add some documentation?

**Comment by jambonrose** [comment]
Thanks for looking at this @timgraham and @loic .

Sorry about the omission of lists/tuples - I've going off of the [documentation](https://docs.djangoproject.com/en/1.8/ref/forms/validation/#raising-validationerror), which states that the use of lists/tuples is "bad," and so I completely overlooked them. Whoops.

Perhaps I'm solving the problem the wrong way. The problem, however, is the following state table:

message | params | output
--------|--------|-------
text    | none   | text
safe    | none   | **safe**
text    | text   | text
safe    | text   | text
text    | safe   | text
safe    | safe   | text

I agree that my solution is not ideal, but I think there is a problem (and perhaps it is with my understanding or the documentation). Notably, the behavior above means many developers are going to opt for the following code, foregoing `params` completely.

```python
raise ValidationError(
    # _ is ugettext
    mark_safe(_(
        'SlugField may not be <code>invalid_value</code>'
    )),
)
```

The problem with the code above is that the invalid value is hard-coded, meaning that if it changes, so must all translations. This violates DRY, and causes the developer and translation teams more work.

By 'interpolate the final message', do you mean the following (untested code):

```python
# _ is ugettext
fmt = _('SlugField may not be {}')
error_string = fmt.format('<code>invalid_value</code>')
raise ValidationError(mark_safe(error_string), code='bad_slug')
```

This solves the problem of DRY and translation, but if the goal is to remove HTML from `ValidationError` instances, the solution above seems undesirable.

The method of overriding `ErrorList/Dict` would require a custom template tag to fully adhere to DRY (to make sure the invalid value only appears in single place). There's something quite elegant about that solution, but it also feels a little over-engineered to me. The real problem is that in this instance, the 'right way' is considerably more complicated than the 'wrong way.'

I'm not sure how to go about solving this, but I'm happy to help if pointed in the right direction.

Thanks again for reading.

**Comment by loic** [comment]
@timgraham, I vote for wontfix, but I wouldn't mind a second opinion. My reasoning is that `ValidationError` shouldn't have to worry about interpolating `SafeData` considering that we support rendering errors in a variety of formats (HTML, JSON, and plain text). Especially considering that interpolating `SafeData` reliably is not trivial (e.g. the proposed interpolation logic is rather fragile).

@jambonrose

> By 'interpolate the final message', do you mean the following (untested code)

Yes that's what I meant, putting HTML inside the error message already restricts it to a specific usage (displaying in a web browser), when you can make this kind of decisions using `params` is not necessary.

That said, if you really want to involve the Django machinery you have a couple of options.

- Do the interpolation in a custom `ErrorList` method (getting the error instances with `self.as_data()` and using their `message` and `params` attributes).

- Or as suggested earlier you can make `SafeText` apply your proposed logic in a way that's transparent to consumers such as `ValidationError`. For example:

```python
import collections

from django.utils.safestring import SafeData, SafeText

class CustomSafeText(SafeText):
    def __mod__(self, other):
        result = super(MySafeText, self).__mod__(other)

        if isinstance(other, tuple):
            values = other
        elif isinstance(other, collections.Mapping):
            values = other.values()
        else:
            values = [other]
        
        if all(isinstance(v, SafeData) for v in values):
            return SafeText(result)
        else:
            return result

# >>> e = ValidationError(CustomSafeText('Hello %(who)s!'), params={'who': SafeText('World')})
# >>> type(list(v)[0])
# django.utils.safestring.SafeText
```

**Comment by timgraham** [comment]
That reasoning makes sense to me.

**Pull request closed without merging**